pkg/solana/logpoller/log_poller.go

```diff
@@ -0,0 +1,150 @@
+package logpoller
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+var (
+	ErrFilterNameConflict = errors.New("filter with such name already exists")
+)
+
+type ORM interface {
+	InsertFilter(ctx context.Context, filter Filter) (id int64, err error)
+	SelectFilters(ctx context.Context) ([]Filter, error)
+	DeleteFilters(ctx context.Context, filters map[int64]Filter) error
+	MarkFilterDeleted(ctx context.Context, id int64) (err error)
+}
+
+type LogPoller struct {
+	services.StateMachine
+	lggr logger.SugaredLogger
+	orm  ORM
+
+	filters *filters
+
+	chStop services.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewLogPoller(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+	lggr = logger.Sugared(logger.Named(lggr, "LogPoller"))
+	return &LogPoller{
+		orm:     orm,
+		lggr:    lggr,
+		filters: newFilters(lggr, orm),
+	}
+}
+
+func (lp *LogPoller) Start(context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		lp.wg.Add(2)
+		go lp.run()
+		go lp.backgroundWorkerRun()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		close(lp.chStop)
+		lp.wg.Wait()
+		return nil
+	})
+}
+
+// RegisterFilter - refer to filters.RegisterFilter for details.
+func (lp *LogPoller) RegisterFilter(ctx context.Context, filter Filter) error {
+	ctx, cancel := lp.chStop.Ctx(ctx)
+	defer cancel()
+	return lp.filters.RegisterFilter(ctx, filter)
+}
+
+// UnregisterFilter refer to filters.UnregisterFilter for details
+func (lp *LogPoller) UnregisterFilter(ctx context.Context, name string) error {
+	ctx, cancel := lp.chStop.Ctx(ctx)
+	defer cancel()
+	return lp.filters.UnregisterFilter(ctx, name)
+}
+
+func (lp *LogPoller) loadFilters(ctx context.Context) error {
+	retryTicker := services.TickerConfig{Initial: 0, JitterPct: services.DefaultJitter}.NewTicker(time.Second)
+	defer retryTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-retryTicker.C:
+		}
+		err := lp.filters.LoadFilters(ctx)
+		if err != nil {
+			lp.lggr.Errorw("Failed loading filters in init logpoller loop, retrying later", "err", err)
+			continue
+		}
+
+		return nil
+	}
+}
+
+func (lp *LogPoller) run() {
```

Comment:
Starting and running go-routines that require a context can be simplified using `services.Engine`.

And example is here: https://github.com/smartcontractkit/chainlink-solana/blob/d60fe71900342811c736459738bcd9c7a6890261/pkg/solana/logpoller/poller.go#L69-L73